### DISCUSSION TITLE
How to specify exact version for port dependency

### DISCUSSION THREAD

**Discussion by lstipakov** [microsoft/vcpkg#24655] [Q&A] [answered]
Hello,

I am maintaining a small out-of-tree port (https://github.com/OpenVPN/openvpn/blob/release/2.5/contrib/vcpkg-ports/pkcs11-helper/portfile.cmake) which has build-time dependency on openssl. Our software, which uses this port, supports both openssl 1.1.1 and 3.0.x and I need a way to specify openssl version I want to use with my port.

When using manifest mode to list dependencies, it could be done like this:

```
{
  "name": "openvpn-gui",
  "version-string": "0.0.1",
  "dependencies": [
    "openssl"
  ],
  "builtin-baseline": "4b766c1cd17205e1b768c4fadfd5f867c1d0510e",
  "overrides": [
    {
      "name": "openssl",
      "version-string": "1.1.1n"
    }
  ]
}
```

In this case the correct openssl version got installed. However, the same approach doesn't work when I use it in port definition (currently port uses CONTROL file, but I am experimenting with vcpkg.json):

```
{
  "name": "pkcs11-helper",
  "version": "1.29.1",
  "description": "pkcs11-helper is a library that simplifies the interaction with PKCS#11 providers for end-user applications.",
  "homepage": "https://github.com/OpenSC/pkcs11-helper",
  "dependencies": [
    "openssl"
  ],
  "builtin-baseline": "4b766c1cd17205e1b768c4fadfd5f867c1d0510e",
  "overrides": [
    {
      "name": "openssl",
      "version-string": "1.1.1n"
    }
  ]
}
```

Here is output, where openssl 3.0.2 is installed instead of 1.1.1n:

```
c:\Users\lev\Projects\vcpkg>vcpkg --overlay-ports=c:\Users\lev\Projects\openvpn\contrib\vcpkg-ports install pkcs11-helper:x64-windows
Computing installation plan...
The following packages will be built and installed:
  * openssl[core]:x64-windows -> 3.0.2#3
    pkcs11-helper[core]:x64-windows -> 1.29.1 -- c:\Users\lev\Projects\openvpn\contrib\vcpkg-ports\pkcs11-helper
Additional packages (*) will be modified to complete this operation.
```

So the question is - how to define exact (or maximum) version for _port_ dependency?

**Comment by autoantwort** [comment, marked as answer]
> So the question is - how to define exact (or maximum) version for port dependency?

Overrides are the right choice, but you can only use them at the root level and not in a port. Within a port you can only use `"version>=": "..."`. So vcpkg uses a [minimum version approach](https://github.com/microsoft/vcpkg/blob/master/docs/specifications/versioning.md#42-declaring-minimum-version-constraints). 

If you could overrides in a port, you could create two port with a dependency on `A` where port one required version 1 and port two requires version 2. This would result in unsolvable version constraints. With the minimum versioning approach you can build any port together without getting versioning conflicts.

**Reply by lstipakov** [reply]
Thank you. I also noticed that documentation (https://vcpkg.readthedocs.io/en/latest/users/versioning/) says:

> Overrides are not considered from ports consumed as a dependency.

It would be nice for such documentation-attentive folks like me to see a warning like "using overrides in ports is not supported".
